Add support for explicit control over model animations

markw65 opened the pull request:
Cesium currently only supports time based animation. This can be inconvenient if the phase of the animation is related to something other than time (eg distance along a path of an object moving at a variable speed).

This came up before in a pull request [here](https://github.com/CesiumGS/cesium/pull/7361), but the author was persuaded that it was better to use nodeTransformations to explicitly control the model. That was (just) doable with that example, because there were just 3 pairs of wheels, all of which needed the exact same, relatively trivial, transformations.

My model has a bicycle chain with 90+ links, two wheels, two chain rings (with different rates of rotation, which are both different again from the rate of rotation of the wheels), two drive arms, and a rider. Using `nodeTransformations` would be extremely complex, and any changes to the model would require updating all the associated javascript. Also, on the site where I use the model, I actually use different models for different types of activity - so I want to just be able to drop in a new model, and have it "just work" (to make it work, I have to ensure that the animation's "duration" in seconds is equal to the distance in meters the model would move during one iteration of the animation).

This adds an `animationTime` property to `ModelAnimation`. If set, it's used by `ModelAnimationCollection.update` to determine the `localAnimationTime`, rather than using the current clock time.

I also added a `manualAnimation` property to `ModelAnimationCollection` so we can still do the short circuit exit from `ModelAnimationCollection.update`.

The new sandcastle example is just a clone of Time Dynamic Wheels, rewritten to use a more complex model, and the new functionality.

cesium-concierge commented:
Thank you so much for the pull request @markw65! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/markw65/cesium/blob/manual-animation/CONTRIBUTORS.md) file!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/manual-animation/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

markw65 commented:
@mramato @hpinkos You reviewed #7361 which originally had a similar intent (although a different approach). Any chance you could take a look? If it is something worth pursuing, any directions on how to write tests for this would be appreciated.

markw65 commented:
@mramato @hpinkos I think this is fully ready for review; I found the existing tests, and modified one to test the new features. I've also updated CONTRIBUTERS.md and CHANGES.md, but don't know how to get cesium-concierge to notice.

markw65 commented:
@OmarShehata maybe you could take a look?

cesium-concierge commented:
Thanks again for your contribution @markw65! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @markw65! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @markw65! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @markw65! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @markw65! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

markw65 commented:
I'm still very interested in getting this merged. I simply have no idea how to "champion" it. Nobody responds to anything I say about it. I'll rebase again, just in case...

ebogo1 commented:
Hi @markw65 - apologies for the long wait and thanks for keeping the PR up to date.

I think adding some kind of manual control over "local time" is a fair extension for animations but the current implementation seems a bit restrictive to me. I could be misunderstanding, but to me it looks like the design around the new `animationTime` property assumes that it'll be used with multiple models updating it similarly (i.e. with the sampled distance property in your new Sandcastle). I wonder if there's a more generalized solution to control the animation time deltas - maybe it'd require a slightly different approach in `ModelAnimationCollection`.

j9liu commented:
Hi @markw65,

I noticed this while working on bringing animations to `ModelExperimental` in #10314. It'd be useful to give users more control over animations!

I feel like the implementation can be reworked somehow. It would be nice to have animations be independent of the scene frame's time, that's for sure. But, at least to me, it doesn't feel right that the user would have to manually increment some value and directly set it in the animation every frame. I'm not sure what would be better API design here, to be honest.

My thoughts aren't very refined, but this is what I'm thinking:
- having some flag to allow animation independent of scene time is not a bad idea, though maybe rename from `manualAnimation` to something else
- `multiplier` is currently a `readonly` property -- what would happen if we allowed the user to set this dynamically?
- if the above suggestion doesn't work, let's rework `animationTime`. Instead of animationTime being a number that is changed, we could make it a property that's some sort of callback. The default would involve the current time (in real life), so that the animation can play even if the scene time pauses, the animation will still continue. But we can also allow this to be overridden by the user.

These are just thoughts, not concrete suggestions, so don't take this as a call to implement them right away. I think having input from other community members and developers would help us figure out where to go from here.

markw65 commented:
Thanks for the feedback!

@ebogo1 
> the design around the new animationTime property assumes that it'll be used with multiple models updating it similarly

I don't really follow? If there are multiple models, they will presumably be moving separately, and will consequently need to have their animationTimes set separately?

eg [this example](https://replayroutes.com/replayroutes.html#clear=1&leader_board=1&wkt=623626f65c38ee99b7f1334bbe95c965&wkt=d6981c606399ab9cb8efe5050c9fc270&autostart=1)

@j9liu 
- I'm not attached to the naming scheme in any way
- multiplier really isn't a good way to go - and this was discussed in #7295, which initially suggested using multiplier for a similar reason. It doesn't work well because you want (eg) the amount the wheels turn by to be proportional to the distance covered by the model. In *theory* you could get that by modifying multiplier, but in practice, with varying speeds, its not easy. In addition, the implementation of multiplier is such that any time you change it, you effectively reset the phase of the animation, so every alteration to multiplier causes a skip in the animation (at least, thats what I remember from following #7295).
- making animationTime a callback which defaults to mimicking the current behavior sounds reasonable. Let me investigate. And if @ebogo1 has any feedback on that suggestion I'd appreciate it...

ebogo1 commented:
I like the idea of turning the new `animationTime` property into something more generic. In this example a sampled distance property works but I think if we're adding more control over the animation system we should keep in mind more generalized use cases. I do agree with @j9liu that setting `animationTime` each tick feels a bit clunky - ideally a callback function that has a generic interface for computing its own local time would be able to handle all sorts of animations.

As a note, I believe [Three.js](https://threejs.org/docs/#api/en/animation/AnimationMixer) and [Babylon.js](https://doc.babylonjs.com/divingDeeper/animation/animation_introduction) both expose a time multiplier as the main way to control animation speed.

markw65 commented:
> As a note, I believe [Three.js](https://threejs.org/docs/#api/en/animation/AnimationMixer) and [Babylon.js](https://doc.babylonjs.com/divingDeeper/animation/animation_introduction) both expose a time multiplier as the main way to control animation speed.

So I agree that multiplier is useful - I'm even using it in the project that uses this patch. Essentially, I use it as a scale factor so that each model would move by `durationSeconds` meters across its full animation cycle, assuming no slipping of wheels.

My claim is that if you want non-time based control over the animation, then multiplier is a very awkward way to go; and the *current* implementation of multiplier simply doesn't work because every time you change it, it causes a jump in where the animation is.

eg suppose multiplier is 1, and you're halfway through the animation cycle, and then you change `multiplier` to 2 (while only allowing time to advance by an insignificant epsilon). Then suddenly, you're at the end of the animation cycle (this assumes you update `_duration` at the same time as `multiplier`; much worse things happen if you don't). You could obviously fix that by also maintaining an `_offset`, and updating that every time `multiplier` changes. But its still not a convenient way to map an arbitrary function onto time...

markw65 commented:
@j9liu 
> The default would involve the current time (in real life), so that the animation can play even if the scene time pauses, the animation will still continue. 

Just wanted to clarify, because as I understand it, thats not the [current behavior](https://github.com/CesiumGS/cesium/blob/main/Source/Scene/ModelAnimationCollection.js#L370). It uses the simulator time, not real time.

j9liu commented:
@markw65 -- You're right, I misspoke. Thanks for the catch!

markw65 commented:
Ok, I've updated the branch. I've added it as a patch on top of the old one, but I think it's probably better to just flatten it; its got more in common with the original code than with my first attempt. Let me know what you'd prefer.

I got rid of manualAnimation, and replaced it with animateWhilePaused. The basic reason for the flag is the same; when its false, we can skip the entire update loop when the scene time hasn't changed (as the current code does). But this time, you only need to set it if you want to provide an animation function that animates while scene time is paused. So unlike the manualAnimation flag, most uses of animationTime won't need to set it (so eg my own use case, although its distance, rather than time based, doesn't need the flag because the distance can't change unless the time does).

I'm not too familiar with the comment based typing scheme; but I think I eventually got it right(?), because my own project is written in typescript, and kept complaining until I got all the comments in place...

markw65 commented:
I think the build failure was a travis glitch (makeZipFile seems to have timed out). I've rebased the branch to get another build going.

markw65 commented:
No, it was a prettier check on my sandcastle example. Should be fixed now.

j9liu commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -46,6 +46,14 @@ function ModelAnimationCollection(model) {
    */
   this.animationRemoved = new Event();
 
+  /**
```
Can we add `@type` and `@default` to the documentation here?

j9liu commented on Source/Scene/ModelAnimation.js:
```diff
@@ -230,4 +240,24 @@ Object.defineProperties(ModelAnimation.prototype, {
+ * A function used to compute the local animation time for a ModelAnimation.
+ * @callback ModelAnimation.AnimationTime
+ *
+ * @param {Number} duration The animation's duration in seconds.
```
```suggestion
 * @param {Number} duration The animation's original duration in seconds.
```

j9liu commented on Source/Scene/ModelAnimation.js:
```diff
@@ -230,4 +240,24 @@ Object.defineProperties(ModelAnimation.prototype, {
+ * @callback ModelAnimation.AnimationTime
+ *
+ * @param {Number} duration The animation's duration in seconds.
+ * @returns {Number} The fractional part of the return value gives
```
This comment seems a bit complicated, but I have to think about how to best rewrite / summarize this. Will come back to it later

j9liu commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -397,22 +409,34 @@ ModelAnimationCollection.prototype.update = function (frameState) {
       duration !== 0.0
-        ? JulianDate.secondsDifference(sceneTime, startTime) / duration
+        ? scheduledAnimation.animationTime
+          ? scheduledAnimation.animationTime(duration)
```
Can we separate these ternary operators so it's more readable? For example

```
let animationDelta= defined(scheduleAnimation.animationTime) ?
   scheduledAnimation.animationTime(duration)
   : JulianDate.secondsDifference(...);
let delta = duration !== 0.0 ? animationDelta : 0.0;
```

markw65 replied:
I took a slightly different approach, because most animationTime functions are going to want to divide by duration, and the default operation *does* divide by duration (you omitted that in your rewrite, but it should have been `JulianDate.secondsDifference(...) / duration`.
Perhaps you meant to write
```
let delta = duration !== 0.0 ? animationDelta / duration : 0.0;
```
but then animationTime has to return something of the form `N * duration + m`, with `N` an integer and `m < duration`. I think its more natural for the animationTime function to return something of the form `N + m` with `N` integer, and `0 <= m < 1`.

j9liu commented on Source/Scene/ModelAnimation.js:
```diff
@@ -230,4 +240,24 @@ Object.defineProperties(ModelAnimation.prototype, {
   },
 });
+/**
+ * A function used to compute the local animation time for a ModelAnimation.
```
Clarify that if this function is defined by the user, the animation will use this instead of the scene time to evaluate its animation.

markw65 replied:
But this is the type definition for the callback. I *do* explicitly clarify that at the point that the animationTime field is defined (at line 42 above).

j9liu commented on Apps/Sandcastle/gallery/Manually Controlled Animation.html:
```diff
@@ -0,0 +1,166 @@
+<!DOCTYPE html>
```
Do you have a model of higher quality that can demonstrate the same functionality for this sandcastle?

j9liu replied:
If you do, can you replace this model and squash the commit history? That way, the addition and removal of the old model won't bloat the size of the change history.

markw65 replied:
I don't have one, but I think I could use the existing "cesium man" model, if that would be better.

j9liu replied:
Cesium Man would work, thank you!

markw65 commented on Source/Scene/ModelAnimation.js:
```diff
@@ -230,4 +240,25 @@ Object.defineProperties(ModelAnimation.prototype, {
+ *          than 1.0
+ *
+ * @example
+ * function animationTime(duration) {
```
Suppose I have several copies of a model, and I want the default animationTime behavior, except that they should all be out of phase with each other. I can just define something like
```
function getAnimationTimeFn(offset) {
  return function(duration, seconds) { return seconds / duration + offset }
}
```

In the same way, it would solve the problem that #10043 is trying to solve (I commented there that the current proposal seems broken).

j9liu replied:
That makes sense, thanks for explaining. I think we should add an example like that to the documentation, as a second `@example` underneath the one already here

markw65 commented on Source/Scene/ModelAnimation.js:
```diff
@@ -230,4 +240,25 @@ Object.defineProperties(ModelAnimation.prototype, {
 });
+/**
+ * A function used to compute the local animation time for a ModelAnimation.
+ * @callback ModelAnimation.AnimationTime
```
will do

markw65 commented:
I think the latest update covers all the feedback/requested changes, except for updating ModelExperimental. I'll take a look at applying the changes there too.

markw65 commented on Specs/Scene/ModelSpec.js:
```diff
@@ -1910,6 +1910,47 @@ describe(
+      const time = JulianDate.fromDate(
+        new Date("January 1, 2014 12:00:00 UTC")
+      );
+      const animations = animBoxesModel.activeAnimations;
```
will do

markw65 commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -46,6 +46,17 @@ function ModelAnimationCollection(model) {
+
   this._model = model;
   this._scheduledAnimations = [];
   this._previousTime = undefined;
```
Actually, would it be better to keep this name, and rename `_prevAnimationTime` to `_prevAnimationDelta` (or some other name). In the original diff _prevAnimationTime was literally the previous value of the animationTime property, but thats no longer the case...

markw65 commented on Source/Scene/ModelAnimationCollection.js:
```diff
@@ -426,7 +440,17 @@ ModelAnimationCollection.prototype.update = function (frameState) {
+
+    if (
+      delta === scheduledAnimation._prevAnimationTime &&
+      !play === (scheduledAnimation._state === ModelAnimationState.STOPPED)
```
I'm changing it to a nested if, to avoid computing `animationStopped` when delta changed (which is presumably the common case, and certainly the case where we care about performance).

markw65 commented:
I've added the animationTime changes to ModelExperimentalAnimation and ModelExperimentalAnimationCollection, added the tests to ModelExperimentalAnimationCollectionSpec.js, and added a second copy of the sandcastle example that uses ModelExperimental.

j9liu commented on Specs/Scene/ModelExperimental/ModelExperimentalAnimationCollectionSpec.js:
```diff
@@ -615,6 +616,100 @@ describe("Scene/ModelExperimental/ModelExperimentalAnimationCollection", functio
+        new Date("January 1, 2014 12:00:00 UTC")
+      );
+      const animationCollection = model.activeAnimations;
+      animationCollection.animateWhilePaused = false;
```
You can remove this line for the `ModelExperimental` spec. The models get deleted after every test here (which is not true for `Model`) so after creation, this will already default to false.

markw65 replied:
This made me realize that I should set animateWhilePaused to false in ModelSpec.js at the end of "animates while paused with an explicit animation time".

So I did, and it broke "animates with a multiplier". The reason being that the first render in "animates with a multiplier" uses the same scene time as all the renders in "animates while paused with an explicit animation time", so the first render relied on having animateWhilePaused set, and after fixing that it took the _previousTime shortcut. So I also changed the render time for "animates while paused with an explicit animation time".

So yes, making each test start fresh is a much better design decision...

j9liu commented on Specs/Scene/ModelExperimental/ModelExperimentalAnimationCollectionSpec.js:
```diff
@@ -615,6 +616,100 @@ describe("Scene/ModelExperimental/ModelExperimentalAnimationCollection", functio
+      scene.renderForSpecs(time);
+      animationTime = 0.1;
+      scene.renderForSpecs(JulianDate.addSeconds(time, 1.0, scratchJulianDate));
+      scene.renderForSpecs(JulianDate.addSeconds(time, 2.0, scratchJulianDate));
```
Add a comment here that this shouldn't trigger an update because the animation time hasn't changed.

j9liu commented on Specs/Scene/ModelExperimental/ModelExperimentalAnimationCollectionSpec.js:
```diff
@@ -615,6 +616,100 @@ describe("Scene/ModelExperimental/ModelExperimentalAnimationCollection", functio
+      scene.renderForSpecs(time);
+      animationTime = 0.1;
+      scene.renderForSpecs(time);
+      scene.renderForSpecs(time);
```
Add a comment here that this shouldn't update the animation because animationTime didn't change

j9liu reviewed (changes requested):
Hi @markw65,

This PR is looking really close, thanks for the quick changes! I left another round of suggestions. Once those are done, I'll do a final pass and look to merge this with main.

markw65 commented on Apps/Sandcastle/gallery/Manually Controlled ModelExperimentalAnimation.html:
```diff
@@ -0,0 +1,167 @@
+<!DOCTYPE html>
```
Will do.

One thing I noticed is that Model.fromGltf takes a "url" option, while ModelExperimental.fromGltf takes a "gltf" option, so (at least on that one point) they're not drop-in compatible. Is that going to be fixed?

j9liu replied:
That's a good point... we might want to add a `url` option for compatibility, then deprecate it after a few months. I'll make a note of it for the future

markw65 commented on Apps/Sandcastle/gallery/Manually Controlled ModelAnimation.html:
```diff
@@ -0,0 +1,167 @@
+<!DOCTYPE html>
```
Will do

j9liu commented on CHANGES.md:
```diff
@@ -10,6 +10,10 @@
 
 - Fixed the inaccurate computation of bounding spheres for `ModelExperimental`. [#10339](https://github.com/CesiumGS/cesium/pull/10339/)
 
+##### Additions :tada:
```
Hm, I'm not sure if there are any bounding sphere issues with `Model`. Does the same bug happen when you render the model with `ModelExperimental`?

j9liu commented:
@markw65 let me know when this is ready for a final review!

j9liu commented:
@markw65 let me know when this is ready for a final review!

markw65 commented:
I think it's ready.

j9liu commented on Apps/Sandcastle/gallery/Manually Controlled Animation.html:
```diff
@@ -0,0 +1,167 @@
+          });
+        });
+        viewer.trackedEntity = modelLabel;
+        modelLabel.viewFrom = new Cesium.Cartesian3(-20.0, -7.0, 4.0);
```
```suggestion
        modelLabel.viewFrom = new Cesium.Cartesian3(-30.0, -10.0, 10.0);
```
Small tweak, these `Cartesian3` values give better visibility of the model when it first loads in.

j9liu commented on Specs/Scene/ModelExperimental/ModelExperimentalAnimationCollectionSpec.js:
```diff
@@ -615,6 +616,102 @@ describe("Scene/ModelExperimental/ModelExperimentalAnimationCollection", functio
+      // no update because scene time didn't change
+      scene.renderForSpecs(JulianDate.addSeconds(time, 2.0, scratchJulianDate));
+      animationTime = 0.3;
+      scene.renderForSpecs(JulianDate.addSeconds(time, 3.0, new JulianDate()));
```
```suggestion
      scene.renderForSpecs(JulianDate.addSeconds(time, 3.0, scratchJulianDate));
```

j9liu reviewed (changes requested):
@markw65 

Looking good! 😄 Just a few tweaks. Once these and the merge conflicts are resolved, I'll go ahead with the merge to main.

markw65 commented:
@j9liu Once I'm done, should I squash it down to a single commit, or leave it as is?

j9liu commented:
@markw65 Leave it as is. Squashing is only important when we need to remove files that were added in the same PR (e.g. the new model and sandcastle).

markw65 commented:
@j9liu Ok, I've applied your suggestions, and rebased to fix CHANGES.md.

My thought on squashing is that it makes the git history cleaner.

j9liu commented:
Merging now -- thanks @markw65 !

markw65 commented:
@j9liu Thank you!
